<launch>
    <!-- SYSTEM MODEL -->
    <!-- parameters -->
    <arg name="device" default="/dev/ttyUSB1" doc="device file of the IMU"/>
    <arg name="baudrate" default="115200" doc="baudrate of the IMU"/>
    <arg name="timeout" default="0.002" doc="timeout for the IMU communication"/>
    <arg name="frame_id" default="/imu" doc="frame id of the IMU"/>
    <arg name="frame_local" default="ENU" doc="desired frame orientation (ENU, NED or NWU)"/>
    <arg name="no_rotation_duration" default="0" doc="duration (int in seconds) of the no-rotation calibration procedure"/>
    <arg name="angular_velocity_covariance_diagonal" default="[0.0004, 0.0004, 0.0004]" doc="Diagonal elements of angular velocity covariance matrix"/>
    <arg name="linear_acceleration_covariance_diagonal" default="[0.0004, 0.0004, 0.0004]" doc="Diagonal elements of linear acceleration covariance matrix"/>
    <arg name="orientation_covariance_diagonal" default="[0.01745, 0.01745, 0.15708]" doc="Diagonal elements of the orientation covariance matrix"/>

    <!-- vehicle parameters -->
    <param name="L_a" type="double" value="0.125" />
    <param name="L_b" type="double" value="0.125" />
    <param name="m" type="double" value="1.98" />
    <param name="I_z" type="double" value="0.03" />
    
    <!-- flags -->
    <param name="sim_flag" type="bool" value="true" />
    <param name="feature_flag" type="bool" value="false" />

    <!-- Track parameters -->
    <param name="trackShape" type="str" value="oval" />
    <!-- <param name="trackShape" type="str" value="3110_big" /> -->
    <!-- <param name="trackShape" type="str" value="L_shape" /> -->

    <!-- Machines -->
    <machine name="odroid_1" address="192.168.10.213" user="odroid" timeout="20.0" env-loader="/home/odroid/barc/env_loader_odroid.sh"/>
    <machine name="odroid_2" address="192.168.10.101" user="odroid" timeout="20.0" env-loader="/home/odroid/barc/env_loader_odroid.sh"/>

    <machine name="this-pc" address="192.168.10.147" user="mpc-ubuntu" timeout="20.0" env-loader="/home/mpcubuntu/barc/env_loader_pc_mpc_ubuntu.sh"/>
    
    <!-- START NODES -->

    <!-- VISUALIZATION -->
    <node machine="this-pc" pkg="barc" type="plotCarTrajectory.py" name="visualization" >
        <param name="plotGPS" type="bool" value="true" />
    </node>

    <!-- AGENT 1 NODES -->

    <!-- LOW LEVEL CONTROLLER -->
    <node machine="odroid_1" pkg="barc" type="controllerLowLevelMultiple.py" name="low_level_controller_1" >
        <param name="car" type="str" value="NewBARC" />
        <remap from="ecu" to="agent_1/ecu"/>
        <remap from="ecu_pwm" to="agent_1/ecu_pwm"/>
    </node>

    <node machine="this-pc" pkg="barc" type="lowLevelPub.py" name="low_level_controller_publisher_1" >
        <param name="car" type="str" value="NewBARC" />
        <remap from="ecu_LMPC" to="agent_1/ecu_LMPC"/>
        <remap from="ecu" to="agent_1/ecu"/>
    </node>

    <!-- Estimator -->
    <node machine="odroid_1" pkg="barc" type="stateEstimatorMultiple.py" name="state_estimator_1" output="screen">
    <!-- delay in state estuiamtor -->
        <param name="beacon" type="int" value="26" />
        <param name="delay_df" type="double" value="0.0" />
        <param name="delay_a" type="double" value="0.0" />
        <!-- ekf properties -->
         <param name="thReset" type="double" value="0.1" />
         <param name="vSwitch" type="double" value="5.5" />
         <param name="psiSwitch" type="double" value="0.5" />

         <param name="Qx_hs" type="double" value="0.01" />
         <param name="Qx_ls" type="double" value="0.01" />

         <param name="Qy_hs" type="double" value="0.01" />
         <param name="Qy_ls" type="double" value="0.01" />

         <param name="Qvx_hs" type="double" value="50.0" />
         <param name="Qvx_ls" type="double" value="50.0" />

         <param name="Qvy_hs" type="double" value="50.0" />
         <param name="Qvy_ls" type="double" value="50.0" />

         <param name="Qax_hs" type="double" value="1.0" />
         <param name="Qax_ls" type="double" value="1.0" />

         <param name="Qay_hs" type="double" value="100.0" />
         <param name="Qay_ls" type="double" value="100.0" />

         <param name="Qpsi_hs" type="double" value="0.1" />
         <param name="Qpsi_ls" type="double" value="0.1" />

         <param name="QpsiDot_hs" type="double" value="1.0" />
         <param name="QpsiDot_ls" type="double" value="0.01" />

         <param name="Rx_hs" type="double" value="1.0" />
         <param name="Rx_ls" type="double" value="1.0" />

         <param name="Ry_hs" type="double" value="1.0" />
         <param name="Ry_ls" type="double" value="1.0" />

         <param name="Rvx_hs" type="double" value="100.01" />
         <param name="Rvx_ls" type="double" value="100.01" />

         <param name="Rax_hs" type="double" value="100.0" />
         <param name="Rax_ls" type="double" value="100.0" />

         <param name="Ray_hs" type="double" value="1.0" />
         <param name="Ray_ls" type="double" value="1.0" />

         <param name="RpsiDot_hs" type="double" value="0.1" />
         <param name="RpsiDot_ls" type="double" value="0.1" />

         <param name="Rvy_hs" type="double" value="100.01" />
         <param name="Rvy_ls" type="double" value="100.01" />

         <remap from="hedge_pos_a" to="agent_1/hedge_pos_a" />
        <remap from="imu/data" to="agent_1/imu/data" />
        <remap from="vel_est" to="agent_1/vel_est" />
        <remap from="ecu" to="agent_1/ecu"/>
        <remap from="pos_info" to="agent_1/pos_info" />
        <remap from="srv_fbk" to="agent_1/srv_fbk"/>

    </node>

    <!-- LMPC CONTROL -->
    <node machine="this-pc" pkg="barc" type="controllerMain.py" name="control_1" output="screen">
        <param name="mode" type="str" value="experiments" />
        <param name="control_mode" type="str" value="TI_MPC"/>
        <param name="saveData" type="bool" value="true" />
        <param name="car" type="str" value="OldBARC" />
        <remap from="pos_info" to="agent_1/pos_info" />
        <remap from="ecu_LMPC" to="agent_1/ecu_LMPC" />
        <remap from="SS" to="agent_1/SS" />
        <remap from="OL_predictions" to="agent_1/OL_predictions" />
        <remap from="other_agent_predictions" to="agent_1/other_agent_predictions" />
    </node>

    <node machine="this-pc" pkg="barc" type="predictor.py" name="predictor" output="screen">
        <param name="this_agent" type="str" value="agent_1"/>
        <param name="other_agent" type="str" value= "agent_2"/>
        <param name="frequency" type="double" value="10" />
    </node>

    <!-- IMU -->
    <node machine="odroid_1" pkg="myahrs_driver" type="myahrs_driver" name="myahrs_driver_1">
        <param name="port" value="/dev/ttyACM0" />
        <param name="baud_rate" value="115200" />
        <remap from="imu/data" to="agent_1/imu/data"/>
    </node>
    <!-- <node machine="odroid" pkg="xsens_driver" type="mtnode.py" name="xsens_driver" output="screen" >
        <param name="device" value="$(arg device)"/>
        <param name="baudrate" value="$(arg baudrate)"/>
        <param name="timeout" value="$(arg timeout)"/>
        <param name="frame_id" value="$(arg frame_id)"/>
        <param name="frame_local" value="$(arg frame_local)"/>
        <param name="no_rotation_duration" value="$(arg no_rotation_duration)"/>
        <rosparam param="angular_velocity_covariance_diagonal" subst_value="True">$(arg angular_velocity_covariance_diagonal)</rosparam>
        <rosparam param="linear_acceleration_covariance_diagonal" subst_value="True">$(arg linear_acceleration_covariance_diagonal)</rosparam>
        <rosparam param="orientation_covariance_diagonal" subst_value="True">$(arg orientation_covariance_diagonal)</rosparam>
        <remap from="imu/data" to="agent_1/imu/data"/>
    </node>  -->

    <!-- GPS -->
    <node machine="odroid_1" pkg="marvelmind_nav" type="hedge_rcv_bin" name="indoor_gps_1" args="/dev/ttyACM1" >
        <remap from="hedge_pos_a" to="agent_1/hedge_pos_a" />
        <remap from="hedge_pos" to="agent_1/hedge_pos"/>
    </node> 
    
    <!-- ARDUINO -->
    <!--  * encoders and ultrasound sensors -->
    <node machine="odroid_1" pkg="rosserial_python" type="serial_node.py" name="arduino_node_1" >
        <param name="port" value="/dev/ttyUSB0" />
        <remap from="vel_est" to="agent_1/vel_est" />
        <remap from="srv_fbk" to="agent_1/srv_fbk"/>
        <remap from="ecu_pwm" to="agent_1/ecu_pwm" />
    </node>

    <!-- AGENT 2 NODEs --> 

    <node machine="odroid_2" pkg="barc" type="controllerLowLevelMultiple.py" name="low_level_controller_2" >
        <param name="car" type="str" value="NewBARC" />
        <remap from="ecu" to="agent_2/ecu"/>
        <remap from="ecu_pwm" to="agent_2/ecu_pwm"/>
    </node>

    <node machine="this-pc" pkg="barc" type="lowLevelPub.py" name="low_level_controller_publisher_2" >
        <param name="car" type="str" value="NewBARC" />
        <remap from="ecu_LMPC" to="agent_2/ecu_LMPC"/>
        <remap from="ecu" to="agent_2/ecu"/>
    </node>

    <!-- Estimator -->
    <node machine="odroid_2" pkg="barc" type="stateEstimatorMultiple.py" name="state_estimator_2" output="screen">
    <!-- delay in state estuiamtor -->
        <param name="beacon" type="int" value="16" />
        <param name="delay_df" type="double" value="0.0" />
        <param name="delay_a" type="double" value="0.0" />
        <!-- ekf properties -->
         <param name="thReset" type="double" value="0.1" />
         <param name="vSwitch" type="double" value="5.5" />
         <param name="psiSwitch" type="double" value="0.5" />

         <param name="Qx_hs" type="double" value="0.01" />
         <param name="Qx_ls" type="double" value="0.01" />

         <param name="Qy_hs" type="double" value="0.01" />
         <param name="Qy_ls" type="double" value="0.01" />

         <param name="Qvx_hs" type="double" value="50.0" />
         <param name="Qvx_ls" type="double" value="50.0" />

         <param name="Qvy_hs" type="double" value="50.0" />
         <param name="Qvy_ls" type="double" value="50.0" />

         <param name="Qax_hs" type="double" value="100.0" />
         <param name="Qax_ls" type="double" value="100.0" />

         <param name="Qay_hs" type="double" value="100.0" />
         <param name="Qay_ls" type="double" value="100.0" />

         <param name="Qpsi_hs" type="double" value="0.1" />
         <param name="Qpsi_ls" type="double" value="0.1" />

         <param name="QpsiDot_hs" type="double" value="1.0" />
         <param name="QpsiDot_ls" type="double" value="0.01" />

         <param name="Rx_hs" type="double" value="1.0" />
         <param name="Rx_ls" type="double" value="1.0" />

         <param name="Ry_hs" type="double" value="1.0" />
         <param name="Ry_ls" type="double" value="1.0" />

         <param name="Rvx_hs" type="double" value="100.01" />
         <param name="Rvx_ls" type="double" value="100.01" />

         <param name="Rax_hs" type="double" value="1.0" />
         <param name="Rax_ls" type="double" value="1.0" />

         <param name="Ray_hs" type="double" value="1.0" />
         <param name="Ray_ls" type="double" value="1.0" />

         <param name="RpsiDot_hs" type="double" value="0.1" />
         <param name="RpsiDot_ls" type="double" value="0.1" />

         <param name="Rvy_hs" type="double" value="100.01" />
         <param name="Rvy_ls" type="double" value="100.01" />

         <remap from="hedge_pos_a" to="agent_2/hedge_pos_a" />
        <remap from="imu/data" to="agent_2/imu/data" />
        <remap from="vel_est" to="agent_2/vel_est" />
        <remap from="ecu" to="agent_2/ecu"/>
        <remap from="pos_info" to="agent_2/pos_info" />
        <remap from="srv_fbk" to="agent_2/srv_fbk"/>

    </node>

    <!-- LMPC CONTROL -->
    <node machine="this-pc" pkg="barc" type="controllerMain.py" name="control_2" output="screen">
        <param name="mode" type="str" value="experiments" />
        <param name="control_mode" type="str" value="TI_MPC"/>
        <param name="saveData" type="bool" value="true" />
        <param name="car" type="str" value="NewBARC" />
        <remap from="pos_info" to="agent_2/pos_info" />
        <remap from="ecu_LMPC" to="agent_2/ecu_LMPC" />
        <remap from="SS" to="agent_2/SS" />
        <remap from="OL_predictions" to="agent_2/OL_predictions" />
        <remap from="other_agent_predictions" to="agent_2/other_agent_predictions" />
    </node>

    <node machine="odroid_2" pkg="xsens_driver" type="mtnode1.py" name="xsens_driver_2" output="screen" >
        <param name="device" value="$(arg device)"/>
        <param name="baudrate" value="$(arg baudrate)"/>
        <param name="timeout" value="$(arg timeout)"/>
        <param name="frame_id" value="$(arg frame_id)"/>
        <param name="frame_local" value="$(arg frame_local)"/>
        <param name="no_rotation_duration" value="$(arg no_rotation_duration)"/>
        <rosparam param="angular_velocity_covariance_diagonal" subst_value="True">$(arg angular_velocity_covariance_diagonal)</rosparam>
        <rosparam param="linear_acceleration_covariance_diagonal" subst_value="True">$(arg linear_acceleration_covariance_diagonal)</rosparam>
        <rosparam param="orientation_covariance_diagonal" subst_value="True">$(arg orientation_covariance_diagonal)</rosparam>
        <remap from="imu/data" to="agent_2/imu/data"/>
    </node> 

    <!-- GPS -->
    <node machine="odroid_2" pkg="marvelmind_nav" type="hedge_rcv_bin" name="indoor_gps_2" args="/dev/ttyACM0" >
        <remap from="hedge_pos_a" to="agent_2/hedge_pos_a" />
        <remap from="hedge_pos" to="agent_2/hedge_pos"/>
    </node> 
    
    <!-- ARDUINO -->
    <!--  * encoders and ultrasound sensors -->
    <node machine="odroid_2" pkg="rosserial_python" type="serial_node.py" name="arduino_node_2" >
        <param name="port" value="/dev/ttyUSB0" />
        <remap from="vel_est" to="agent_2/vel_est" />
        <remap from="srv_fbk" to="agent_2/srv_fbk"/>
        <remap from="ecu_pwm" to="agent_2/ecu_pwm" />
    </node>


</launch>
